File: app/src/main/java/com/toshi/crypto/HDWallet.java
/*
 * 	Copyright (c) 2017. Toshi Inc
 *
 * 	This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.toshi.crypto;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.toshi.crypto.hdshim.EthereumKeyChainGroup;
import com.toshi.crypto.keyStore.KeyStoreHandler;
import com.toshi.crypto.util.TypeConverter;
import com.toshi.exception.InvalidMasterSeedException;
import com.toshi.exception.KeyStoreException;
import com.toshi.util.FileNames;
import com.toshi.util.LogUtil;
import com.toshi.view.BaseApplication;

import org.bitcoinj.core.NetworkParameters;
import org.bitcoinj.crypto.DeterministicKey;
import org.bitcoinj.crypto.MnemonicException;
import org.bitcoinj.wallet.DeterministicSeed;
import org.bitcoinj.wallet.KeyChain;
import org.bitcoinj.wallet.UnreadableWalletException;
import org.bitcoinj.wallet.Wallet;

import java.io.IOException;

import rx.Single;

import static com.toshi.crypto.util.HashUtil.sha3;

public class HDWallet {

    private static final String ALIAS = "MasterSeedAlias";
    private static final String MASTER_SEED = "ms";

    private SharedPreferences prefs;
    private ECKey identityKey;
    private ECKey paymentKey;
    private String masterSeed;

    public HDWallet() {
        this.prefs = BaseApplication.get().getSharedPreferences(FileNames.WALLET_PREFS, Context.MODE_PRIVATE);
    }

    public HDWallet(@NonNull final SharedPreferences preferences) {
        this.prefs = preferences;
    }

    public Single<HDWallet> getExistingWallet() {
        return Single.fromCallable(() -> {
            this.masterSeed = readMasterSeedFromStorage();
            if (this.masterSeed == null) throw new InvalidMasterSeedException(new Throwable("Master seed is null"));
            final Wallet wallet = initFromMasterSeed(this.masterSeed);
            deriveKeysFromWallet(wallet);

            return this;
        });
    }

    // WARNING: This will delete any wallet stored on disk
    public Single<HDWallet> createWallet() {
        return Single.fromCallable(() -> {
            final Wallet wallet = generateNewWallet();
            deriveKeysFromWallet(wallet);

            return this;
        });
    }

    private Wallet generateNewWallet() {
        final Wallet walletForSeed = new Wallet(getNetworkParameters());
        final DeterministicSeed seed = walletForSeed.getKeyChainSeed();
        final Wallet wallet = new Wallet(getNetworkParameters(),  new EthereumKeyChainGroup(getNetworkParameters(), seed));
        final String masterSeed = seedToString(seed);
        saveMasterSeedToStorage(masterSeed);
        return wallet;
    }

    private Wallet initFromMasterSeed(final String masterSeed) {
        try {
            final DeterministicSeed seed = getSeed(masterSeed);
            seed.check();
            return constructFromSeed(seed);
        } catch (final UnreadableWalletException | MnemonicException e) {
            throw new RuntimeException("Unable to create wallet. Seed is invalid");
        }
    }

    public Single<HDWallet> createFromMasterSeed(final String masterSeed) {
        return Single.fromCallable(() -> {
            try {
                final DeterministicSeed seed = getSeed(masterSeed);
                seed.check();
                final Wallet wallet = constructFromSeed(seed);
                deriveKeysFromWallet(wallet);
                saveMasterSeedToStorage(masterSeed);
                return this;
            } catch (final UnreadableWalletException | MnemonicException e) {
                throw new InvalidMasterSeedException(e);
            }
        });
    }

    @NonNull
    private Wallet constructFromSeed(final DeterministicSeed seed) {
        return new Wallet(getNetworkParameters(),  new EthereumKeyChainGroup(getNetworkParameters(), seed));
    }

    private DeterministicSeed getSeed(final String masterSeed) throws UnreadableWalletException {
        return new DeterministicSeed(masterSeed, null, "", 0);
    }

    private NetworkParameters getNetworkParameters() {
        return NetworkParameters.fromID(NetworkParameters.ID_MAINNET);
    }

    private void deriveKeysFromWallet(final Wallet wallet) {
        try {
            deriveIdentityKey(wallet);
            derivePaymentKey(wallet);
        } catch (final UnreadableWalletException | IOException ex) {
            throw new RuntimeException("Error deriving keys: " + ex);
        }
    }

    private void deriveIdentityKey(final Wallet wallet) throws IOException, UnreadableWalletException {
        this.identityKey = deriveKeyFromWallet(wallet, 0, KeyChain.KeyPurpose.AUTHENTICATION);
    }

    private void derivePaymentKey(final Wallet wallet) throws IOException, UnreadableWalletException {
        this.paymentKey = deriveKeyFromWallet(wallet, 0, KeyChain.KeyPurpose.RECEIVE_FUNDS);
    }

    private ECKey deriveKeyFromWallet(final Wallet wallet, final int iteration, final KeyChain.KeyPurpose keyPurpose) throws UnreadableWalletException, IOException {
        DeterministicKey key = null;
        for (int i = 0; i <= iteration; i++) {
            key = wallet.freshKey(keyPurpose);
        }

        if (key == null) {
            throw new IOException("Unable to derive key");
        }

        return ECKey.fromPrivate(key.getPrivKey());
    }

    public String signIdentity(final String data) {
        return sign(data.getBytes(), this.identityKey);
    }

    public String signTransaction(final String data) {
        try {
            final byte[] transactionBytes = TypeConverter.StringHexToByteArray(data);
            return sign(transactionBytes, this.paymentKey);
        } catch (final Exception e) {
            LogUtil.print(getClass(), "Unable to sign transaction. " + e);
            return null;
        }
    }

    public String signTransactionWithoutMinus27(final String data) {
        try {
            final byte[] transactionBytes = TypeConverter.StringHexToByteArray(data);
            return signWithoutMinus27(transactionBytes, this.paymentKey);
        } catch (final Exception e) {
            LogUtil.print(getClass(), "Unable to sign transaction. " + e);
            return null;
        }
    }

    public String hashAndSignTransactionWithoutMinus27(final String data) {
        try {
            final byte[] transactionBytes = TypeConverter.StringHexToByteArray(data);
            return hashAndSignWithoutMinus27(transactionBytes, this.paymentKey);
        } catch (final Exception e) {
            LogUtil.print(getClass(), "Unable to sign transaction. " + e);
            return null;
        }
    }

    private String sign(final byte[] bytes, final ECKey key) {
        return sign(bytes, key, true);
    }

    private String sign(final byte[] bytes, final ECKey key, final Boolean doSha3Hash) {
        final byte[] msgHash = doSha3Hash ? sha3(bytes) : bytes;
        final ECKey.ECDSASignature signature = key.sign(msgHash);
        return signature.toHex();
    }

    private String signWithoutMinus27(final byte[] bytes, final ECKey key) {
        final ECKey.ECDSASignature signature = key.sign(bytes);
        return signature.toHexWithNoMinus27();
    }

    private String hashAndSignWithoutMinus27(final byte[] bytes, final ECKey key) {
        final byte[] msgHash = sha3(bytes);
        final ECKey.ECDSASignature signature = key.sign(msgHash);
        return signature.toHexWithNoMinus27();
    }

    public String getMasterSeed() {
        return this.masterSeed;
    }

    public byte[] generateDatabaseEncryptionKey() {
        final byte[] encryptionKey = new byte[64];
        final byte[] privateKey = getPrivateKeyBytes();
        System.arraycopy(privateKey, 0, encryptionKey, 0, 32);
        System.arraycopy(privateKey, 0, encryptionKey, 32, 32);
        return encryptionKey;
    }

    private byte[] getPrivateKeyBytes() {
        if (this.identityKey != null) {
            return this.identityKey.getPrivKeyBytes();
        }
        return null;
    }

    public String getOwnerAddress() {
        if (this.identityKey == null) return null;
        return TypeConverter.toJsonHex(this.identityKey.getAddress());
    }

    public String getPaymentAddress() {
        if(this.paymentKey != null) {
            return TypeConverter.toJsonHex(this.paymentKey.getAddress());
        }
        return null;
    }

    @Override
    public String toString() {
        return "Identity: " + getOwnerAddress() + "\nPayment: " + getPaymentAddress();
    }

    private void saveMasterSeedToStorage(final String masterSeed) {
        try {
            final KeyStoreHandler keyStoreHandler = new KeyStoreHandler(BaseApplication.get(), ALIAS);
            final String encryptedMasterSeed = keyStoreHandler.encrypt(masterSeed);
            saveMasterSeed(encryptedMasterSeed);
            this.masterSeed = masterSeed;
        } catch (KeyStoreException e) {
            throw new IllegalStateException(e);
        }
    }

    private void saveMasterSeed(final String masterSeed) {
        this.prefs.edit()
                .putString(MASTER_SEED, masterSeed)
                .apply();
    }

    private String readMasterSeedFromStorage() {
        try {
            final KeyStoreHandler keyStoreHandler = new KeyStoreHandler(BaseApplication.get(), ALIAS);
            final String encryptedMasterSeed = this.prefs.getString(MASTER_SEED, null);
            if (encryptedMasterSeed == null) return null;
            return keyStoreHandler.decrypt(encryptedMasterSeed, this::saveMasterSeed);
        } catch (KeyStoreException e) {
            throw new IllegalStateException(e);
        }
    }

    public void clear() {
        this.prefs
                .edit()
                .clear()
                .apply();
    }

    private String seedToString(final DeterministicSeed seed) {
        if (seed == null || seed.getMnemonicCode() == null) {
            return null;
        }

        final StringBuilder sb = new StringBuilder();
        for (final String word : seed.getMnemonicCode()) {
            sb.append(word).append(" ");
        }

        // Remove the extraneous space character
        sb.deleteCharAt(sb.length() - 1);

        return sb.toString();
    }
}

File: app/src/main/java/com/toshi/crypto/HDWallet.kt
/*
 * 	Copyright (c) 2017. Toshi Inc
 *
 * 	This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.toshi.crypto

import com.toshi.crypto.util.HashUtil.sha3
import com.toshi.crypto.util.TypeConverter
import com.toshi.exception.InvalidKeySetException
import com.toshi.exception.SignTransactionException
import com.toshi.util.logging.LogUtil
import com.toshi.util.sharedPrefs.WalletPrefsInterface
import rx.Observable
import rx.Scheduler
import rx.Single
import rx.schedulers.Schedulers
import rx.subjects.BehaviorSubject
import java.util.concurrent.Executors

class HDWallet(
        private val walletPrefs: WalletPrefsInterface,
        private val identityKey: ECKey,
        private val paymentKeys: List<ECKey>,
        val masterSeed: String,
        private val scheduler: Scheduler = Schedulers.from(Executors.newSingleThreadExecutor())
) {
    private val paymentAddressSubject = BehaviorSubject.create<String>()
    private val nameSubject = BehaviorSubject.create<String>()

    val paymentAddress: String
        get() {
            return paymentAddressSubject.value
                    ?: throw IllegalStateException("PaymentAddress is null")
        }

    val name: String
        get() {
            return nameSubject.value
                    ?: throw IllegalStateException("Wallet name is null")
        }

    val ownerAddress: String
        get() {
            val identityAddress = identityKey.address
            return TypeConverter.toJsonHex(identityAddress)
        }

    fun getPaymentAddressObservable(): Observable<String> = paymentAddressSubject.asObservable()

    fun getWalletNameObservable(): Observable<String> = nameSubject.asObservable()

    init {
        if (paymentKeys.isEmpty()) throw InvalidKeySetException("No payment keys in list")
        val currentWalletIndex = getCurrentWalletIndex()
        publishCurrentPaymentAddress(currentWalletIndex)
        publishCurrentWalletName(currentWalletIndex)
    }

    fun signIdentity(data: String): String {
        try {
            return sign(data.toByteArray(), identityKey)
        } catch (e: Exception) {
            LogUtil.exception("Unable to sign identity. $e")
            throw SignTransactionException("Unable to sign identity")
        }
    }

    fun signTransaction(data: String): Single<String> {
        return Single.fromCallable {
            val currentWalletIndex = getCurrentWalletIndex()
            val key = getKeyFromIndex(currentWalletIndex)
            return@fromCallable signDataWithPaymentKey(key, data)
        }
        .subscribeOn(scheduler)
    }

    private fun signDataWithPaymentKey(paymentKey: ECKey, data: String): String {
        try {
            val transactionBytes = TypeConverter.StringHexToByteArray(data)
            return sign(transactionBytes, paymentKey)
        } catch (e: Exception) {
            LogUtil.exception("Unable to sign transaction", e)
            throw SignTransactionException("Unable to sign transaction")
        }
    }

    fun signTransaction(data: String, hash: Boolean): Single<String> {
        return Single.fromCallable {
            val bytes = TypeConverter.StringHexToByteArray(data)
            val transactionBytes = if (hash) sha3(bytes) else bytes
            val key = getKeyFromIndex(getCurrentWalletIndex())
            return@fromCallable signWithoutMinus27(transactionBytes, key)
        }
        .subscribeOn(scheduler)
    }

    private fun sign(bytes: ByteArray, key: ECKey, doSha3Hash: Boolean = true): String {
        val msgHash = if (doSha3Hash) sha3(bytes) else bytes
        val signature = key.sign(msgHash)
        return signature.toHex()
    }

    private fun signWithoutMinus27(bytes: ByteArray, key: ECKey): String {
        val signature = key.sign(bytes)
        return signature.toHexWithNoMinus27()
    }

    fun generateDatabaseEncryptionKey(): ByteArray {
        val encryptionKey = ByteArray(64)
        val privateKey = identityKey.privKeyBytes ?: throw IllegalStateException("Bad identity key")
        System.arraycopy(privateKey, 0, encryptionKey, 0, 32)
        System.arraycopy(privateKey, 0, encryptionKey, 32, 32)
        return encryptionKey
    }

    fun changeWallet(index: Int): Int {
        if (paymentKeys.size <= index) throw IllegalStateException("Index is bigger than paymentKeys size")
        saveCurrentIndexToPrefs(index)
        publishCurrentPaymentAddress(index)
        publishCurrentWalletName(index)
        return index
    }

    private fun publishCurrentPaymentAddress(index: Int) {
        val address = addressFromIndex(index)
        paymentAddressSubject.onNext(address)
    }

    private fun addressFromIndex(index: Int): String {
        val key = getSafeKeyFromIndex(index)
        return TypeConverter.toJsonHex(key.address)
    }

    private fun publishCurrentWalletName(index: Int) {
        val name = getWalletNameFromIndex(index)
        nameSubject.onNext(name)
    }

    fun getAddressesWithNames(): List<Pair<String, String>> {
        return getAddresses()
                .mapIndexed { index, address -> Pair(address, getWalletNameFromIndex(index)) }
    }

    private fun getWalletNameFromIndex(index: Int) = "Wallet ${index + 1}"

    @Throws(IllegalStateException::class)
    private fun getKeyFromIndex(index: Int): ECKey = paymentKeys.getOrNull(index)
            ?: throw IllegalStateException("Couldn't get key from index: $index")

    private fun getSafeKeyFromIndex(index: Int): ECKey = paymentKeys.getOrNull(index) ?: paymentKeys[0]

    private fun saveCurrentIndexToPrefs(index: Int) = walletPrefs.setCurrentWalletIndex(index)

    fun getCurrentWalletIndex() = walletPrefs.getCurrentWalletIndex()

    fun clear() = walletPrefs.clear()

    fun hasAddress(paymentAddress: String): Boolean {
        val walletAddresses = getAddresses()
        return walletAddresses.contains(paymentAddress)
    }

    fun getAddresses(): List<String> = paymentKeys.map { TypeConverter.toJsonHex(it.address) }

    override fun toString(): String {
        val identityAddress = identityKey.address
        return "Identity: $identityAddress"
    }
}
